### PR TITLE
#7630: Keyboard vertical navigation in the text lines next to objects should move caret to position closest to object

### LINKED ISSUE #7630: Keyboard vertical navigation in text lines next to isLimit elements that are not isInline
## 📝 Provide a description of the improvement

This is currently locally implemented inside table cells. There is a custom key handler that checks if caret is in first/last line of text (visually) and if navigation would cross `isLimit` boundary then it's altered to move caret to beginning/end of that limit.

The above could be extended globally so also handling caret next to block widgets (not only on the last text position as it is handled currently) and arrows inside other nested editables like image caption.

---

### **Navigating across widgets using the keyboard**

**Case**: `p + widget + widget + p`

**Expected**:

1.  A selection in the last line of a text (for instance `<paragraph>foo[]bar</paragraph><widget></widget>)`
2.  <kbd>Arrow down</kbd>: Caret jumps to last text position that is before widget.
3.  <kbd>Arrow down</kbd>: Fake caret of widget-type-around feature get's activated above widget edge _(WTA is handling this currently)._
4.  <kbd>Arrow down</kbd>: The first widget gets selected _(WTA is handling this currently)_.
5.  <kbd>Arrow down</kbd>: Fake caret gets activated between the widgets  _(WTA is handling this currently)_.
6.  <kbd>Arrow down</kbd>: The second widget gets selected _(WTA is handling this currently)_.
7.  <kbd>Arrow down</kbd>: Fake caret of widget-type-around feature get's activated below widget edge _(WTA is handling this currently)._
8.  <kbd>Arrow down</kbd>: The selection lands in the last paragraph.

**Nice to have**:

1.  <kbd>Shift</kbd>+<kbd>Arrow</kbd> – above should be aware of <kbd>Shift</kbd> key and extend selection if it's pressed.

## 📃 Other details

*   Browser: …
*   OS: …
*   CKEditor version: …
*   Installed CKEditor plugins: …

---

If you'd like to see this improvement implemented, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by niegowski**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Feature (widget): Keyboard vertical navigation in the text lines next to objects should move caret to position closest to object. Closes #7630.

Internal (table): Keyboard vertical navigation inside a table cell extracted to the widget plugin. 

---

### Additional information

### CODE REVIEW

**Comment by niegowski** [comment]
@ckeditor/qa-team guys could you check if it works correctly? This is <kbd>up</kbd>/<kbd>down</kbd> arrow handling extracted from inside table cells to more generic cases of approaching any limit element (so edges of nested editables when caret is inside it and approaching block widgets from outside) with <kbd>up</kbd>/<kbd>down</kbd> arrows.

This is not covering <kbd>Shift</kbd>+<kbd>Arrows</kbd>

**Comment by FilipTokarski** [comment]
Looks ok. 👍

**Comment by jodator** [comment]
Quick request: as this goes to the widget feature I think that it would be nice to have a dedicated manual test.

Now, how the test will be created might be different (or not) to the regular manual tests that you've worked on lately (like in tables). Please take a look at other manual tests, like UI, engine/view, utils to get ideas.

What I mean by that is that the manual test for this might not require a fully running editor. My guess here is that it _might_ be possible to create a manual test for navigation in some stripped form _but_ I can be wrong.

To sum up:

*   Manual test is needed.
*   Check the best way to test this feature (might not require full editor).

File: packages/ckeditor5-widget/tests/verticalnavigation.js [hunk 1]
```diff
@@ -0,0 +1,1379 @@
...
+					border: 0 !important
+				}
+				.ck-editor__editable:not(.ck-editor__nested-editable ) {
+					width: 300px !important;
+				}
+				td { width: 30px !important; }
+				tr:nth-child(2) td:nth-child(2) { width: 300px !important; }
+				`
+			) );
+			global.document.querySelector( 'head' ).appendChild( styleElement );
```
Comment (jodator) [inline packages/ckeditor5-widget/tests/verticalnavigation.js:155]: Some explanation is needed - this might be a major WTF if someone doesn't know the test problems. I gues is that you need to control dimensions to have tests repeatable.

Also, did you check if there is no other way (Karma/mocha plugin, etc)?

Another question - is it dependant on the font available/used?

File: packages/ckeditor5-widget/tests/verticalnavigation.js [hunk 2]
```diff
@@ -0,0 +1,1379 @@
...
+		} );
+
+		describe( 'simple cell text content', () => {
+			describe( 'with collapsed selection', () => {
+				beforeEach( () => {
+					setModelData( model, modelTable( [
+						[ '00', '01', '02' ],
+						[ '10', 'foo[]bar', '12' ],
+						[ '20', '21', '22' ]
+					] ) );
```
Comment (jodator) [inline packages/ckeditor5-widget/tests/verticalnavigation.js:523]: I think that a general (defined ad-hoc) nested editable would be better than a specialised table construct.

something like:

```html
<paragraph>Foo bar</paragraph>
<editable>before</editable>
<editable>foo[]bar</editable>
<editable>after</editable>
```

or, if needed the `<editable>` might be two-level deep (like image): `<block><editable></editable></block>`.

The reason for this is that not everybody must know tables internals and it's nuisances to work with a general widget feature.

Especially because the test suite is for "with selection inside nested content editable" but sub-suite is called "simple cell text content".

File: packages/ckeditor5-widget/src/verticalnavigation.js [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+		const range = model.createRange( startPosition, endPosition );
+		const lastRangePosition = getNearestTextPosition( model.schema, range, 'backward' );
+
+		if ( lastRangePosition && startPosition.isBefore( lastRangePosition ) ) {
+			return model.createRange( startPosition, lastRangePosition );
+		}
+
+		return null;
+	} else {
```
[the remark this answers is not in the document]
Comment (niegowski) [reply packages/ckeditor5-widget/src/verticalnavigation.js:100]: I prefer it this way because it clearly shows opposite direction handling :-)

File: packages/ckeditor5-widget/src/verticalnavigation.js [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+		// Navigation is in the opposite direction than the selection direction so this is shrinking of the selection.
+		// Selection for sure will not approach any object.
+		if ( expandSelection && !selection.isCollapsed && selection.isBackward == arrowDownPressed ) {
+			return;
+		}
+
+		// Find a range between selection and closest limit element.
+		const range = findTextRangeFromSelection( editing, selection, arrowDownPressed );
+
+		if ( !range || range.start.isTouching( range.end ) ) {
```
[the remark this answers is not in the document]
Comment (niegowski) [reply packages/ckeditor5-widget/src/verticalnavigation.js:41]: TBH I can't remember why. After changing to `range.isCollapsed` all tests are still passing so maybe it was some old use case that is no longer valid.

File: packages/ckeditor5-widget/tests/verticalnavigation.js (hunk 1 above)
Comment (niegowski) [reply packages/ckeditor5-widget/tests/verticalnavigation.js:155]: This is strictly moved from test in tables (tests of the same feature)

Comment (niegowski) [reply packages/ckeditor5-widget/tests/verticalnavigation.js:155]: > Another question - is it dependant on the font available/used?

Unfortunately yes - that's why gecko is excluded from some of those tests.

File: packages/ckeditor5-table/src/tablekeyboard.js [hunk 5]
```diff
@@ -219,46 +218,17 @@ export default class TableKeyboard extends Plugin {
...
-					newSelection.setFocus( newPosition );
+		const cellRange = model.createRangeIn( tableCell );
 
-					writer.setSelection( newSelection );
-				} else {
-					writer.setSelection( newPosition );
-				}
-			} );
+		// Let's check if the selection is at the beginning/end of the cell.
+		if ( this._isSelectionAtCellEdge( selection, cellRange, isForward ) ) {
```
[the remark this answers is not in the document]
Comment (niegowski) [reply packages/ckeditor5-table/src/tablekeyboard.js]: Added

File: packages/ckeditor5-table/src/tablekeyboard.js [hunk 6]
```diff
@@ -219,46 +218,17 @@ export default class TableKeyboard extends Plugin {
...
-		// We can't move the selection directly to the another cell because of dual position at the end/beginning
-		// of wrapped line (it's at the same time at the end of one line and at the start of the next line).
-		if ( this._isSingleLineRange( textRange, isForward ) ) {
-			model.change( writer => {
-				const newPosition = isForward ? cellRange.end : cellRange.start;
-
-				if ( expandSelection ) {
-					const newSelection = model.createSelection( selection.anchor );
-					newSelection.setFocus( newPosition );
+		const cellRange = model.createRangeIn( tableCell );
```
[the remark this answers is not in the document]
Comment (niegowski) [reply packages/ckeditor5-table/src/tablekeyboard.js]: Nice catch :) Fixed.

File: packages/ckeditor5-table/src/tablekeyboard.js [hunk 7]
```diff
@@ -281,7 +252,9 @@ export default class TableKeyboard extends Plugin {
 		// If the current limit element is not table cell we are for sure not at the cell edge.
 		// Also `modifySelection` will not let us out of it.
 		if ( !schema.getLimitElement( focus ).is( 'element', 'tableCell' ) ) {
-			return false;
+			const boundaryPosition = isForward ? cellRange.end : cellRange.start;
```
[the remark this answers is not in the document]
Comment (niegowski) [reply packages/ckeditor5-table/src/tablekeyboard.js]: :)

File: packages/ckeditor5-widget/src/verticalnavigation.js [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+		const expandSelection = data.shiftKey;
+		const selection = model.document.selection;
+
+		if ( !arrowUpPressed && !arrowDownPressed ) {
+			return;
+		}
+
+		// Navigation is in the opposite direction than the selection direction so this is shrinking of the selection.
+		// Selection for sure will not approach any object.
+		if ( expandSelection && !selection.isCollapsed && selection.isBackward == arrowDownPressed ) {
```
[the remark this answers is not in the document]
Comment (niegowski) [reply packages/ckeditor5-widget/src/verticalnavigation.js]: Extracted to helper function.

File: packages/ckeditor5-widget/src/verticalnavigation.js [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+	const schema = model.schema;
+	const range = model.createRangeIn( startPosition.root );
+
+	for ( const { previousPosition, item, type } of range.getWalker( { startPosition, direction } ) ) {
+		if ( schema.isLimit( item ) && !schema.isInline( item ) ) {
+			return previousPosition;
+		}
+
+		// Stop looking for isLimit element if the next element is a block element (it is for sure not single line).
+		if ( type == ( direction == 'forward' ? 'elementStart' : 'elementEnd' ) && schema.isBlock( item ) ) {
```
[the remark this answers is not in the document]
Comment (niegowski) [reply packages/ckeditor5-widget/src/verticalnavigation.js]: Local variable extracted.

File: packages/ckeditor5-widget/tests/verticalnavigation.js (hunk 2 above)
Comment (niegowski) [reply packages/ckeditor5-widget/tests/verticalnavigation.js:523]: I was to lazy. I'll change this :)

File: packages/ckeditor5-widget/tests/verticalnavigation.js (hunk 1 above)
Comment (jodator) [reply packages/ckeditor5-widget/tests/verticalnavigation.js:155]: Only this thing left - please, add a comment why this is needed.

Comment (jodator) [reply packages/ckeditor5-widget/tests/verticalnavigation.js:155]: Side note, as PHPStorm was highlighting this: the `style.type` is not needed: https://wiki.developer.mozilla.org/en-US/docs/Web/HTML/Element/style#Attributes.
